* CARD GAME PROBABILITIES AND STATS

The calculation of the probability of an event "A" is the following:

· P(A) = Event outcomes favourable to A / Sample space

The probability is always between 0 and 1.

There are 52 cards in a standard deck, among those 52 cards there are 4 Aces. 
The probability of drawing and Ace is calculated as follows: P(A) = 4/52.

In [5]:
# Sample Space
cards = 52

# Outcomes
aces = 4

# Divide possible outcomes by the sample set
ace_probability = aces / cards

# Print probability rounded to two decimal places
print(round(ace_probability, 2))

# Ace Probability Percent Code
ace_probability_percent = ace_probability * 100

# Print probability percent rounded to one decimal place
print(str(round(ace_probability_percent, 0)) + '%')

0.08
8.0%


More examples. Let's calculate the probability of drawing a card that is a Heart, a Face card (such as Jacks, Queens, or Kings), or a combination of both, such as the Queen of Hearts. In such cases, I want to create a User-Defined Function (UDF) event_probability() to which I pass the event_outcomes and the sample_space to find the probability of an event in percentage form.

In [6]:
# Create function that returns probability percent rounded to two decimal place
def event_probability(event_outcomes, sample_space):
    probability = (event_outcomes / sample_space) * 100
    return round(probability, 2)

# Sample Space
cards = 52

# Determine the probability of drawing a heart
hearts = 13
heart_probability = event_probability(hearts, cards)

# Determine the probability of drawing a face card
face_cards = 12
face_card_probability = event_probability(face_cards, cards)

# Determine the probability of drawing the queen of hearts
queen_of_hearts = 1
queen_of_hearts_probability = event_probability(queen_of_hearts, cards)

# Print each probability
print(str(heart_probability) + '%')
print(str(face_card_probability) + '%')
print(str(queen_of_hearts_probability) + '%')

25.0%
23.08%
1.92%


- Probability with Combinations and Permutations.

Determining the size of your sample space is key to calculating probabilities. Fortunately, there are ways to make the counting task easier.
Two of these ways are permutations and combinations.

- Permutations

Permutations are the number of ways a subset of a specified size can be arranged from a given set, generally without replacement.

The formula to calculate permutations is the following:

nPk = n! / (n-k)!

Where: 

  n = the full set of numbers to choose from.

  k = the numbers that you will actually choose.

To find the number of permutations of pocket Aces, you'll consider the full set of aces to choose from (n=4) and you also consider the number of aces that you actually choose (k=2).

4P2 = 4! / (4-2)!

- Combinations

When you're working with permutations, the order matters. With combinations, however, this isn't the case: the order doesn't matter.
This means that if you want to figure out how many combinations you actually have, you just create all the permutations and divide by all the redundancies.

Combinations refers to the number of ways a subset of a specified size can be drawn from a given set.

The formula to calculate permutations is the following: 

nCk = nPk! / k!

Where:

  nPk = permutations.

  k = the number of cards that you will actually choose.

Consider another example with Aces. There are four Aces in a deck of cards, and these are all the different combinations of pocket Aces;

1 · Ace Hearts & Ace Diamonds

2 · Ace Hearts & Ace Clubs

3 · Ace Hearts & Ace Spades

4 · Ace Diamonds & Ace Clubs

5 · Ace Diamonds & Ace Spades

6 · Ace Clubs & Ace Spades

There are six combinations of pocket Aces. To find the number of combinations, you first must find the number of permutations.

In [7]:
# Permutations Code
import math
n = 4
k = 2

# Determine permutations and print result
Permutations = math.factorial(n) / math.factorial(k)
print(Permutations)

12.0


In [8]:
# Combinations Code
n = 52
k = 2

# Determine Permutations
Permutations = math.factorial(n) / math.factorial(n - k)

# Determine Combinations and print result
Combinations = Permutations / math.factorial(k)
print(Combinations)

1326.0


* Independent Vs. Dependent Events

· Independent events are events that don't impact the probability of the other event(s).

· Two events A and B are independent if knowing whether event A occurred gives no information about whether event B occurred.

· Dependent events, then, are events that have an impact on the probability of the other event(s).

· For example, you draw a card from the deck and then draw a second card from the deck without replacing the first card.

· In this case, the probability of drawing an Ace the fist draw is not the same as the probability of drawing an Ace on the second draw.

· Let's consider the example where you already know the probability of drawing an Ace on the first draw.

· Now you need to determine the probability of drawing an Ace on the second draw, if the first card drawn was either a King or an Ace:


In [9]:
# Sample Space
cards = 52
cards_drawn = 1 
cards = cards - cards_drawn 

# Determine the probability of drawing an Ace after drawing a King on the first draw
aces = 4
ace_probability1 = event_probability(aces, cards)

# Determine the probability of drawing an Ace after drawing an Ace on the first draw
aces_drawn = 1
aces = aces - aces_drawn
ace_probability2 = event_probability(aces, cards)

# Print each probability
print(ace_probability1)
print(ace_probability2)

7.84
5.88


* Texas Hold'em Poker

The game is played with a standard 52 card deck. 

Each player is dealt two cards to start the hand and will make the best five-card hand possible by using their two cards combined with the five community cards that are dealt throughout the hand. 

Cards are dealt in four rounds:

· Pre-Flop: Each player is dealt two cards, known as "hole cards".

· Flop: Three community cards are dealt.

· Turn: One community card is dealt.

· River: Final community card is dealt.

Example: 

· Your hand: Ace of diamonds and Queen of Diamonds.

· Situation: You are on the turn. Meaning, already four cards were dealt, and only one card is pending to be dealt ("the river" card).

· Community cards: Jack of clubs, 9 of diamonds, 4 of diamonds, 2 of spades.

· You have a Flush "draw" meaning if the river card turns out to be a Diamond you will have five diamonds, a strong hand, called "Flush" in Poker terminology.

What's the probability that the last community card is a Diamond?

In [10]:
# Sample Space
cards = 52
hole_cards = 2
turn_community_cards = 4
cards = cards - (hole_cards + turn_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4
# In poker, cards that complete a draw are known as "outs"
outs = diamonds - diamonds_drawn

#Determine river flush probability
river_flush_probability = event_probability(outs, cards)
print(river_flush_probability)


19.57


Example: 

· Your hand: Jack and Ten of Hearts.

· Situation: You are on the turn. Meaning, already four cards were dealt, and only one card is pending to be dealt ("the river" card).

· Community cards: 9 of diamonds, 3 of clubs, Queen of spades and 5 of hearts.

· You have a Straight "draw" open ended, meaning if the river card turns out to be a King or an Eight you will complete a straight.

What's the probability that the River card completes the Straight?

In [11]:
# Sample Space
cards = 52
hole_cards = 2
turn_community_cards = 4
cards = cards - (hole_cards + turn_community_cards)

# Outcomes
eights = 4
kings = 4
outs = eights + kings

# Determine river straight probability
river_straight_probability = event_probability(outs, cards)
print(river_straight_probability)

17.39


- Multiple Events

· To calculate the probability for multiple events, you basically determine the number of events, you then determine the probability for each event occurring separately and you multiply all of these probabilities to get your final answer.

- Mutually Exclusive Events

· When you're working with multiple events, you might also have events that are mutually exclusive: they cannot both occur. In such cases, you might want to calculate the probability of any of multiple mutually exclusive events occurring. In such cases, you don't multiply probabilities, but you simply sum together the probability of each event occurring.

- Examples

Determine the probability of the following mutually exclusive events.

· Drawing a heart or drawing a club.

· Drawing an ace, a king or a queen.

In [12]:
# Sample Space
cards = 52

# Calculate the probability of drawing a heart or a club
hearts = 13
clubs = 13
heart_or_club = event_probability(hearts, cards) + event_probability(clubs, cards)

# Calculate the probability of drawing an ace, king, or a queen
aces = 4
kings = 4
queens = 4
ace_king_or_queen = event_probability(aces, cards) + event_probability(kings, cards) + event_probability(queens, cards)

print(heart_or_club)
print(ace_king_or_queen)

50.0
23.07


- Non-Mutually Exclusive Events

· Not all events are mutually exclusive: Drawing a diamond or drawing an ace are two non-mutually exclusive events.

· The ace of diamonds is both an ace and a diamond. When events are not mutually exclusive, you must correct for the overlap.

- Examples

Determine the probability of the following mutually exclusive events.

· Drawing a heart or drawing an ace.

· Drawing a red card or a face card.

In [13]:
# Sample Space
cards = 52

# Calculate the probability of drawing a heart or an ace
hearts = 13
aces = 4
ace_of_hearts = 1
heart_or_ace = event_probability(hearts, cards) + event_probability(aces, cards) - event_probability(ace_of_hearts, cards)

# Calculate the probability of drawing a red card or a face card
red_cards = 26
face_cards = 12
red_face_cards = 6
red_or_face_cards = event_probability(red_cards, cards) + event_probability(face_cards, cards) - event_probability(red_face_cards, cards)

print(round(heart_or_ace, 1))
print(round(red_or_face_cards, 1))

30.8
61.5


- Intersection of Independent Events

· The probability of the intersection of two independent events is determined by multiplying the probabilities of each event occurring.

In [14]:
# Sample Space
cards = 52

# Outcomes
aces = 4

# Probability of one ace
ace_probability = aces / cards

# Probability of two consecutive independant aces 
two_aces_probability = ace_probability * ace_probability

# Two Ace Probability Percent Code
two_ace_probability_percent = two_aces_probability * 100
print(round(two_ace_probability_percent, 1))

0.6


- Intersection of Dependent Events

The probability of the intersection of two non-independent events (Event A & Event B given A) is determined by multiplying the probability of Event A occurring, times the probability of Event B given A.

· What is the probability of being dealt two Aces?


In [15]:
# Sample Space first draw
cards = 52

# Outcomes first draw
aces = 4

# Probability of ace on first draw
first_ace_probability = aces / cards

# Sample Space second draw
cards = cards - 1

# Outcomes second draw
aces = aces - 1

# Probability of ace on second draw after ace on first
second_ace_probability = aces / cards

# Probability of two consecutive aces (dependent)
both_aces_probability = first_ace_probability * second_ace_probability * 100
print(both_aces_probability)

0.4524886877828055


- Intersection of Dependent Events: Flop Flush Draw

Example: 

· Your hand: Ace and Queen of Diamonds.

· Situation: You are on flop.

· Community cards: Jack of clubs, 9 of diamonds, 4 of diamons.

· You have a Flush "draw" , meaning if the turn or river card turns out to be a diamond the flush will be completed.

What's the probability of getting a Flush by the river?

These are all the different possible scenarios (each scenario is mutually-exclusive, in other words, one of these four scenarios is definitely going to occur):

A) Turn Diamond, River Non Diamond (Made Flush)

B) Turn Non Diamond, River Diamond (Made Flush)

C) Turn Diamond, River Diamond (Made Flush)

D) Turn Non Diamond, River Non Diamond (No Flush)

· The simplest approach to figure this out is to determine the probability of scenario D, and subtract that from 1.

In [16]:
# Sample Space on turn
cards = 52
hole_cards = 2
flop_community_cards = 3
cards = cards - (hole_cards + flop_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4
non_diamonds_drawn = 1
outs = diamonds - diamonds_drawn
turn_non_diamonds = cards - outs - non_diamonds_drawn 

# Probability of not getting a diamond on the turn
no_diamond_turn_probability = turn_non_diamonds / cards

# Sample Space on river
turn_community_card = 1
cards = cards - turn_community_card

# Outcomes on river
river_non_diamonds = turn_non_diamonds - turn_community_card

# Probability of not getting a diamond on the river
no_diamond_river_probability = river_non_diamonds / cards

# Probability of not getting a flush
no_flush_probability = no_diamond_turn_probability * no_diamond_river_probability

# Probability of getting a flush
flush_probability = 1 - no_flush_probability
flush_probability_percent = flush_probability * 100

# Print probability percent rounded to one decimal place
print(round(flush_probability_percent, 1))

38.4


- Example: 

· Your hand: 6 and 5 of Diamonds.

· Situation: You are on flop.

· Community cards: Jack of clubs, 9 of diamonds, 4 of diamons.

· You still have a flush draw, but this time you don’t have the Ace.

· If a Diamond falls on the turn and the river, there is a good chance someone will have a better flush draw. 

Determine the probability of a Diamond falling on the Turn or the River, but not both.

· You already have the probability of not hitting a flush.

· Now find the probability of a diamond falling on the turn and river, add that to the probability of not hitting a flush, and subtract from 1.

In [17]:
# Sample Space on turn
cards = 52
hole_cards = 2
flop_community_cards = 3
cards = cards - (hole_cards + flop_community_cards)

# Outcomes on turn
diamonds = 13
diamonds_drawn = 4
outs = diamonds - diamonds_drawn

# Probability of diamond on turn
diamond_turn_probability = outs / cards

# Sample Space on river
turn_diamonds = 1
cards = cards - turn_diamonds

# Outcomes on river
river_diamonds = outs - turn_diamonds

# Probability of diamond on river
diamond_river_probability = river_diamonds / cards

# Probability of getting a diamond on the turn and river
two_diamonds_probability = diamond_turn_probability * diamond_river_probability

# Determine the probability of getting only one diamond by the river
one_diamond_probability = 1 - (no_flush_probability + two_diamonds_probability)
one_diamond_probability_percent = one_diamond_probability * 100

# Print probability percent rounded to one decimal place
print(round(one_diamond_probability_percent, 1))

35.1


- EXPECTED VALUE

· "How much do I gain - or lose - on average, if I repeatedly play this game?"

· If the possible outcomes of the game and their associated probabilities can be described by a random variable, then you can answer the above question by computing its expected value, which is equal to a weighted average of the outcomes where each outcome is weighted by its probability.

· EV = Value x Probability

What is the expected value if there is $100 (Value) in the pot, and your probability of winning the pot is 0.75?

In [18]:
# Initialize `pot` and `probability` variables
pot = 100
probability = 0.75

# Determine expected value
expected_value = pot * probability
print(expected_value)

75.0


- Example: 

· Your hand: Ace and Queen of Diamonds.

· Opponent's hand: 2 of hearts and 2 of clubs.

· Situation: You are on the turn.

· Community cards: Jack of clubs, 9 of diamonds, 4 of diamonds, 2 of clubs.

· You have a flush draw, and you have the Ace.

· Total pot: $60

· Opponents bet: $20

To win the hand on the River, you must hit any Diamond except a Jack or 2.

The Jack or 2 of Diamonds would give your opponent a better hand, a full house and four of a kind respectively.

You have to call $20 to stay in the hand, and if you win the hand you win 60. 

· If your expected value is greater than $20 you should call the bet, and if not you should fold.

In [19]:
# Sample Space
cards = 52
hole_cards = 2
# Your opponent provided you information... use it!
opponents_hole_cards = 2 

turn_community_cards = 4
cards = cards - (hole_cards + opponents_hole_cards + turn_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4

# You can't count the two diamonds that won't help you win
diamond_non_outs = 2 

outs = diamonds - diamonds_drawn - diamond_non_outs

# Determine win probability
win_probability = outs / cards

# Determine expected value
pot = 60
ev = pot * win_probability

# Print ev and appropriate decision
call_amount = 20
if ev >= 20:
    print(round(ev, 2), 'Call')
else:
    print(round(ev, 2), 'Fold')

9.55 Fold


- Your EV is only $9.55.

- You would need to risk 20.00 dollars to get the 9.55 dollars reward, so you must fold.

- Determining EV is critical to understand at the Poker table.